src/v/cloud_storage/async_manifest_view.cc

```diff
@@ -1401,62 +968,120 @@ async_manifest_view::hydrate_manifest(
-            "archive_start_offset: {}, cached: {}",
-            _stm_manifest.get_archive_start_offset(),
-            _last_archive_start_offset));
-    }
-
-    // Perform simple scan of the manifest list.
-    vlog(
-      _ctxlog.trace,
-      "Scanning manifests list with {} elements, query: {}",
-      _manifests->components.size(),
-      to_string(query));
-    auto it = std::find_if(
-      _manifests->components.begin(),
-      _manifests->components.end(),
-      [this, query](const spillover_manifest_path_components& c) {
+    const auto& manifests = _stm_manifest.get_spillover_map();
+    auto ix = ss::visit(
+      query,
+      [&](model::offset o) {
+          vlog(
+            _ctxlog.debug,
+            "search_spillover_manifest query: {}, num manifests: {}, first: "
+            "{}, last: {}",
+            query,
+            manifests.size(),
+            manifests.empty() ? model::offset{}
+                              : manifests.begin()->base_offset,
+            manifests.empty() ? model::offset{}
+                              : manifests.last_segment()->committed_offset);
+          const auto& bo_col = manifests.get_base_offset_column();
+          const auto& co_col = manifests.get_committed_offset_column();
+          auto co_it = co_col.lower_bound(o);
+          if (co_it.is_end()) {
+              return -1;
+          }
+          auto bo_it = bo_col.at_index(co_it.index());
+          while (!bo_it.is_end()) {
+              if (o >= *bo_it && o <= *co_it) {
+                  return static_cast<int>(bo_it.index());
+              }
+              ++bo_it;
+              ++co_it;
+          }
+          return -1;
+      },
+      [&](kafka::offset k) {
           vlog(
-            _ctxlog.trace,
-            "Scanning manifest with components {}, query: {}",
-            c,
-            to_string(query));
-          return contains(c, query);
+            _ctxlog.debug,
+            "search_spillover_manifest query: {}, num manifests: {}, first: "
+            "{}, last: {}",
+            query,
+            manifests.size(),
+            manifests.empty() ? kafka::offset{}
+                              : manifests.begin()->base_kafka_offset(),
+            manifests.empty() ? kafka::offset{}
+                              : manifests.last_segment()->next_kafka_offset());
+          const auto& bo_col = manifests.get_base_offset_column();
+          const auto& co_col = manifests.get_committed_offset_column();
+          const auto& do_col = manifests.get_delta_offset_column();
+          const auto& de_col = manifests.get_delta_offset_end_column();
+          auto bo_it = bo_col.begin();
+          auto co_it = co_col.begin();
+          auto do_it = do_col.begin();
+          auto de_it = de_col.begin();
+          while (!bo_it.is_end()) {
+              auto bko = kafka::offset(*bo_it - *do_it);
+              auto nko = kafka::offset(*co_it - *de_it);
```

Comment:
The previous version of the search landed in `segment_meta::next_kafka_offset` and `segment_meta::bas_kafka_offset` which defaults the delta offset to 0 if it's equal to its default value of `model::offset_delta::min`. Should we replicate that here?

Reply:
done